### PR TITLE
Add message bubbles

### PR DESCRIPTION
**Pull request by su-ex**
Related to  vector-im/element-web#4635

<details>
<summary>Screenshots</summary>

</details>

Signed-off-by: Quirin Götz <codeworks@supercable.onl>

### CODE REVIEW

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:552]:
```suggestion
                { this.renderThemeSection() }
                { SettingsStore.getValue("feature_new_layout_switcher") ? this.renderLayoutSection() : null }
                { this.renderFontSection() }
                { this.renderAdvancedSection() }
```

File: src/components/views/rooms/MessageComposer.tsx
Comment (SimonBrandner) [inline src/components/views/rooms/MessageComposer.tsx:446]: I don't think I understand the purpose of these comments

File: src/components/views/messages/TextualBody.js
Comment (SimonBrandner) [inline src/components/views/messages/TextualBody.js:540]:
```suggestion
                    <span className="mx_MEmoteBody mx_EventTile_content" dir="auto">
```

File: src/components/views/messages/TextualBody.js
Comment (SimonBrandner) [inline src/components/views/messages/TextualBody.js:556]:
```suggestion
                    <span className="mx_MNoticeBody mx_EventTile_content" dir="auto">
```

File: src/components/views/messages/TextualBody.js
Comment (SimonBrandner) [inline src/components/views/messages/TextualBody.js:564]:
```suggestion
                    <span className="mx_MTextBody mx_EventTile_content" dir="auto">
```

File: src/components/views/messages/TextualBody.js
Comment (SimonBrandner) [inline src/components/views/messages/TextualBody.js:62]: We usually don't use prefixes in prop names

File: src/components/views/messages/MessageEvent.js
Comment (SimonBrandner) [inline src/components/views/messages/MessageEvent.js:55]: We usually don't use prefixes in prop names

File: res/css/views/rooms/_BubbleLayout.scss
Comment (SimonBrandner) [inline res/css/views/rooms/_BubbleLayout.scss:22]: Should we be using an `sc` prefix?

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
I didn't have the time to go through everything but generally, this looks awesome! Just some minor things and questions. Thank you! :rocket: 

The stuff with the prefixes is probably up to the team.

And feel free to scream, if my suggestions don't make sense :sweat_smile: 

Edit: Also, regarding the curly braces... we've settled on `{ foo, bar }`, instead of `{foo, bar}` for objects but a rule for the `react-sdk` hasn't been setup yet

File: res/css/views/rooms/_BubbleLayout.scss
Comment (su-ex) [reply res/css/views/rooms/_BubbleLayout.scss:22]: In one place this will collide if not.

File: src/components/views/messages/TextualBody.js
Comment (su-ex) [reply src/components/views/messages/TextualBody.js:564]: I think I removed this by intention because it broke rtl (either Chromium, Firefox or both) on one occasion. I've added `dir="auto"` on another place therefore and this should work in both. Actually I've tested the rtl stuff thoroughly and it worked better with the bubble layout and the normal layout.

File: src/i18n/strings/hu.json
Comment (su-ex) [reply src/i18n/strings/hu.json:3376]: One check complained about the strings, is that why? Just throw them away then and keep only en_EN?

Comment (SimonBrandner) [reply src/i18n/strings/hu.json:3376]: Developers shouldn't change any of the files by hand. You can run `yarn i18n` to generate the English file and commit it and the rest should be left to Weblate

File: src/components/views/messages/MVideoBody.tsx
Comment (su-ex) [reply src/components/views/messages/MVideoBody.tsx:33]: This should preferably stay searchable what it would if just removing sc because things makes maintenance a whole lot easier. Is just removing sc what you mean?

File: src/components/views/elements/EventListSummary.tsx
Comment (su-ex) [reply src/components/views/elements/EventListSummary.tsx:70]: Copy'n'paste, I assume. 😅

File: src/components/views/messages/MImageBody.js
Comment (su-ex) [reply src/components/views/messages/MImageBody.js:50]: `scBubble` could maybe become `bubbleEnabled`. It's not widely used yet anyway.

File: src/i18n/strings/hu.json
Comment (su-ex) [reply src/i18n/strings/hu.json:3376]: Hmm, ok.

File: src/components/views/rooms/MessageComposer.tsx
Comment (su-ex) [reply src/components/views/rooms/MessageComposer.tsx:446]: The IRC layout uses the group layout for the message composer because it has no own rules for that, the bubble layout has. I left them in sperated to easily be able to adapt this if that changes.

File: src/components/views/messages/TextualBody.js
Comment (su-ex) [reply src/components/views/messages/TextualBody.js:564]: Oops, ignore, weird line break on mobile, didn't see that's not about `dir="auto"`. 🤪

File: src/components/views/messages/MVideoBody.tsx
Comment (SimonBrandner) [reply src/components/views/messages/MVideoBody.tsx:33]: > Is just removing sc what you mean?

Yes

File: src/components/views/rooms/MessageComposer.tsx
Comment (SimonBrandner) [reply src/components/views/rooms/MessageComposer.tsx:446]: Ah, sorry. I more of meant why is there the commented code? 

```
// "mx_IRCLayout": this.props.layout == Layout.IRC,
// "mx_GroupLayout": this.props.layout == Layout.Group,
```

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:490]:
```suggestion
                        disabled={this.state.layout !== Layout.Bubble || this.state.adaptiveSideBubbles}
```

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:499]:
```suggestion
                        disabled={this.state.layout !== Layout.Bubble}
```

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:498]:
```suggestion
                        onChange={(checked) => this.setState({ adaptiveSideBubbles: checked })}
```

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:499]: Also, does it make sense to also disable this when `singleSideBubbles` is enabled?

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:252]:
```suggestion
        this.setState({ layout: layout });
```

File: src/components/views/rooms/MessageComposer.tsx
Comment (su-ex) [reply src/components/views/rooms/MessageComposer.tsx:446]: If the irc layout gets its own rules for the message composer those lines can be commented in and the line after those two removed. It was just for remeberance once that might happen.

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:391]:
```suggestion
            <span className="mx_SettingsTab_subheading">{ _t("Message layout") }</span>
```

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (SimonBrandner) [inline src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:411]:
```suggestion
                        { "IRC" }
```

File: src/contexts/RoomContext.ts
Comment (SimonBrandner) [inline src/contexts/RoomContext.ts:44]: I am probably totally overlooking where this is used

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
This would probably be for the second part of the review. I haven't made myself review the `EventTile.tsx` and `_BubbleLayout.scss` changes yet, but I think we should probably first decide on the things like the prefixes

File: src/components/views/rooms/MessageComposer.tsx
Comment (SimonBrandner) [reply src/components/views/rooms/MessageComposer.tsx:446]: Ah, ok. I think that if this changes at any time, it probably won't be a big problem to add these lines, so it's not necessary to keep them there. If you're keen on keeping it, something like this might be better:
```suggestion
				// When IRC layout gets something for the message composer we can use the following
                // "mx_IRCLayout": this.props.layout == Layout.IRC,
                // "mx_GroupLayout": this.props.layout == Layout.Group,

                // IRC layout has nothing for message composer so use group layout stuff
```

File: res/css/views/rooms/_BubbleLayout.scss
Comment (SimonBrandner) [reply res/css/views/rooms/_BubbleLayout.scss:22]: **For people that weren't part of the discussion on Matrix:** After discussing we've decided it would be better to use `mx_` and avoid conflicts by some class renaming

File: src/components/views/elements/EventListSummary.tsx
Comment (SimonBrandner) [inline src/components/views/elements/EventListSummary.tsx:84]:
```suggestion
        if (layout === Layout.Bubble) {
```

for consistency

File: src/components/views/rooms/MessageComposer.tsx
Comment (su-ex) [reply src/components/views/rooms/MessageComposer.tsx:446]: Whatever, I don't really care. 🙃

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (su-ex) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:490]: For consistency one `=`. 😁

File: src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx
Comment (su-ex) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:499]: No. If a default config enables both you might get stuck.

Comment (su-ex) [reply src/components/views/settings/tabs/user/AppearanceUserSettingsTab.tsx:499]: `adaptiveSideBubbles` sort of overrides `singleSideBubbles`. That's why it's also fine semantically, I think.

File: src/components/views/elements/EventListSummary.tsx
Comment (su-ex) [reply src/components/views/elements/EventListSummary.tsx:84]: Weren't somewhere a change three to two? :thinking: 
So three is what should be used?

Comment (SimonBrandner) [reply src/components/views/elements/EventListSummary.tsx:84]: Generally, both are used all over the codebase, though I think we should use `===` if we can. 

> Weren't somewhere a change three to two?

Do you mean one of my suggestions? If so, I don't thinks

File: src/contexts/RoomContext.ts
Comment (su-ex) [reply src/contexts/RoomContext.ts:44]: Once a build failed if `singleSideBubbles` wasn't there. Recently I noticed that `adaptiveSideBubbles` was missing and added it just in case. But I have absolutely no clue why this is needed either.

Comment (SimonBrandner) [reply src/contexts/RoomContext.ts:44]: _Odd_

File: src/components/views/elements/EventListSummary.tsx
Comment (su-ex) [reply src/components/views/elements/EventListSummary.tsx:84]: Yeah, that's what I thought. Then I've misseen it.

File: src/components/structures/RoomView.tsx
Comment (SimonBrandner) [inline src/components/structures/RoomView.tsx:1675]:
```suggestion
            for (const messageList of messagelists) {
                const boundingBox = messagelist.getBoundingClientRect();
                if (boundingBox.width > width) width = boundingBox.width;
            }
```

This mainly a personal preference, so feel free to ignore

File: src/components/views/rooms/EventTile.tsx
Comment (SimonBrandner) [inline src/components/views/rooms/EventTile.tsx:681]:
```suggestion
            if (this.props.layout === Layout.Bubble) {
```

File: src/components/views/rooms/EventTile.tsx
Comment (SimonBrandner) [inline src/components/views/rooms/EventTile.tsx:717]:
```suggestion
            if (this.props.layout === Layout.Bubble) {
```

File: src/components/views/rooms/EventTile.tsx
Comment (SimonBrandner) [inline src/components/views/rooms/EventTile.tsx:908]:
```suggestion
            mx_EventTile_info: isInfoMessage && this.props.layout !== Layout.Bubble,
```

**Review by SimonBrandner — APPROVED** [review]
LGTM otherwise. I have to admit that I've mostly skimmed through the CSS and the `EventTile` code. I am starting to understand why reviews can take a while sometimes :sweat_smile:

File: src/components/structures/RoomView.tsx
Comment (su-ex) [reply src/components/structures/RoomView.tsx:1675]: I'm not sure this is the optimal solution anyway. Here I'm iterating over the "normal" message list and the one of the search. Thus the iteration. I feel this can be solved better (more Reacty) with some sort of references or the like. However, I couldn't figure out a working solution although I struggled quite some time with this.
I'd like some input here. **I** definitely can't solve that in a better way than I did.

Comment (SimonBrandner) [reply src/components/structures/RoomView.tsx:1675]: From a quick look, I can't see an easy way to fix this either. I'll try to think about it some more though I am not sure it has a simple solution

File: src/components/structures/RoomView.tsx
Comment (SimonBrandner) [inline src/components/structures/RoomView.tsx:625]: Are we calling this on every update? Won't this cause a loop? What am I missing? :sweat_smile:

Comment (su-ex) [reply src/components/structures/RoomView.tsx:625]: It doesn't create a loop. 🤷
It might make sense to split the width detection part out into another function and call it from `onResize()` and here though ...

Comment (SimonBrandner) [reply src/components/structures/RoomView.tsx:625]: Yeah, you're right, it doesn't. I assume this is because of `shouldComponentUpdate()` which I didn't realize was there. You can mark this as resolved because GH doesn't let me do it...

File: src/components/structures/RoomView.tsx
Comment (SimonBrandner) [reply src/components/structures/RoomView.tsx:1675]: Sorry, why is it that we can't just use refs?

Comment (su-ex) [reply src/components/structures/RoomView.tsx:1675]: I didn't get them working. I need one to the search results panel's and the normal message list.

Comment (SimonBrandner) [reply src/components/structures/RoomView.tsx:1675]: Ah, I'll try to play with them a bit too, if I get time, to see if I run into the same problem

Comment (su-ex) [reply src/components/structures/RoomView.tsx:1675]: I'm pretty sure it _can_ be solved better for with refs or the like.

**Review by SimonBrandner — APPROVED** [review]

**Comment by turt2live** [comment]
hey @su-ex can we get some screenshots in the PR description to make review easier? Thanks!

**Comment by turt2live** [comment]
@SimonBrandner now that your name is on the commits as well, we'll need signoff from you on here as well please.

**Comment by SimonBrandner** [comment]
> @SimonBrandner now that your name is on the commits as well, we'll need signoff from you on here as well please.

Ah, true:
```
Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>
```

**Comment by su-ex** [comment]
@turt2live Sure, added! :camera:

**Comment by nadonomy** [comment]
@su-ex thanks for cherry picking your commits & iterating on your implementation into this PR after the discussion we had on Matrix, really appreciated!

The PR principally looks good, but we need to do a bit of internal planning (with @gsouquet @amshakal & @jryans) on precisely how we'll iterate on this so will figure out some details with them before we approve/merge.

Relatedly, we wouldn't consider https://github.com/vector-im/element-web/issues/4635 closed until we're closer to shipping this on production, including de-risking some cross-platform efforts, so I've edited your PR description to avoid this.

Thanks again for the effort!

**Comment by turt2live** [comment]
Thank you for the work on this, and sparking the idea of how it could work - we really appreciate it. We've picked up the idea over at https://github.com/matrix-org/matrix-react-sdk/pull/6291 and hope to land it soon from the looks of it.

In the meantime, closing this in favour of https://github.com/matrix-org/matrix-react-sdk/pull/6291 - thanks again for taking a stab at this.

**Pull request closed without merging**
